jupyterlab/jupyterlab
# [WIP] Automated UI testing using Galata

- Adds UI testing support using https://github.com/jupyterlab/galata
- Tests are run as part of CI with `ui-tests` workflow added
- After tests are run, a result report is generated and uploaded. It can be accessed in GitHub Actions page. If any of the tests fail, image diffs and failing call-stack can be inspected from report. Simply download the artifact and open `test/report/index.html` for the interactive report.
- Detailed documentation for adding new UI tests is added in `ui-tests/README.md`
- Helper npm scripts and a shell script added to help local development and testing. Usage described in the documentation as well.

## Test Plan
- [x] General UI Tests:
  - Launch screen, toggling Simple Mode, toggling Dark Theme
  - Open and capture main menu and sub menus
  - Open and capture sidebar tabs
  - Move sidebar panel to right
- [x] Notebook Create
  - Create cells for each cell type (raw, markdown, code)
  - Save and capture notebook
  - Toggle Dark Theme
- [x] Notebook Run
  - Upload notebook and image dependency to JupyterLab
  - Check File Browser list for uploaded files
  - Run notebook and capture notebook panel
  - Check numerical cell output values
- [x] Cell editing
  - Create a notebook with couple of cells
  - Modify notebook using operations in Edit menu (cut, copy, paste, delete, select, deselect, move, split, merge)
- [x] Notebook Toolbar
  - Create a notebook with couple of cells
  - Run actions in notebook toolbar
- [x] Table of Contents
  - Create tags
  - Create a notebook with markdown and code cells
  - Select cells and assign tags
  - Show TOC panel, filter using cell types and tags

Below is the capture from a test run:
![jlab-galata](https://user-images.githubusercontent.com/40003442/120088229-6f091300-c0a3-11eb-9cdb-630d3bed4ebe.gif)
